File: 1-arrays&strings/exercises/merge_ranges.py
def merge_ranges(meetings):

    # Sort by starttime
    sorted_meetings = sorted(meetings)

    # Initialize merged_meetings with the earliest meeting
    merged_meetings = [sorted_meetings[0]]

    for current_meeting_start, current_meeting_end in sorted_meetings[1:]:
        last_merged_meeting_start, last_merged_meeting_end = merged_meetings[-1]

        # If the current meeting overlaps with the last merged meeting, use the
        # later end time of the two
        if (current_meeting_start <= last_merged_meeting_end):
            merged_meetings[-1] = (last_merged_meeting_start,
                                   max(last_merged_meeting_end,
                                       current_meeting_end))
        else:
            # Add the current meeting since it doesn't overlap
            merged_meetings.append((current_meeting_start, current_meeting_end))

    return merged_meetings

if __name__ == "__main__":
    print(merge_ranges([(0, 1), (3, 5), (4, 8), (10, 12), (9, 10)]))

File: 4-search&sort/exercises/merge_ranges.go
package main

import "sort"

type Times struct {
	Start int
	End   int
}

func mergeRanges(list []Times) {

	sort.SliceStable(list, func(i, j int) bool {
		return list[i].End < list[j].End
	})
	mergedMeetings := []Times{list[0]}
	for _, current := range list {
		last := mergedMeetings[len(list)-1]
		if current.Start <= last.Start {
			mergedMeetings[len(list)-1] = Times{
				Start: last.Start,
				End:   max(last.End, current.End),
			}
		} else {
			mergedMeetings = append(mergedMeetings, Times{
				Start: current.Start,
				End:   current.End,
			})
		}
	}
}

// Max returns the larger of x or y.
func max(x int, y int) int {
	if x < y {
		return y
	}
	return x
}
